File: Lab/exer01/sabile_ex01.py
# Jerico Sabile
# CSMC180: Introduction to Parallel Computing
# Exer01: Computing the Pearson Correlation Coefficient of a Matrix and a Vector
# Feb. 5, 2024 - 12:00

import time
import random

def pearson_cor(x, y, m, n):
    v = []
    for i in range(n):
        v.append((m*sumXY(x,y,m,i)-sumX(x,m,i)*sumY(y,m))/((m*sumX2(x,m,i)-sumX(x,m,i)**2)*((m*sumY2(y,m))-sumY(y,m)**2))**0.5)
    return v

def sumY2(y,m):
    sum = 0
    for i in range(m): sum += y[i]**2
    return sum

def sumX2(x,m,j):
    sum = 0
    for i in range(m): sum += x[i][j]**2
    return sum

def sumXY(x,y,m,j):
    sum = 0
    for i in range(m): sum += x[i][j] * y[i] 
    return sum

def sumX(x,m,j):
    sum = 0
    for i in range(m): sum += x[i][j]
    return sum

def sumY(y,m):
    sum = 0
    for i in range(m): sum += y[i]
    return sum

def generateRandom(x, y):
    numbers = []
    for i in range(x):
        if y == 0: numbers.append(random.randint(0,100))
        else: 
            new = [] 
            for j in range(y): new.append(random.randint(0,100))
            numbers.append(new)
    return numbers 

# TEST CASE
# ANSWER: 0.47
# matrix = [[3.63],[3.02],[3.82],[3.42],[3.59],[2.87],[3.03],[3.46],[3.36],[3.3]]
# y = [53.1,49.7,48.4,54.2,54.9,43.7,47.2,45.2,54.4,50.4]
# m = 10
# n = 1

size = int(input("Size: "))
matrix = generateRandom(size, size)
y = generateRandom(size, 0)
m = n = size

start_time = time.time()
ans = pearson_cor(matrix, y, m, n)
time_elapsed = time.time() - start_time
print(f"time elapsed: {time_elapsed} seconds")
# print(f"r={ans}")

File: Lab/exer01/sabile_ex01.c
#include <stdio.h>
#include <stdlib.h>
#define __USE_GNU
#include <sched.h>
#include <errno.h>
#include <unistd.h>
#include <pthread.h>
#include <math.h>
#include <time.h>

long sumY2(int* y, int m) {
    long sum = 0;
    for(int i=0; i<m; i++) sum += pow(y[i], 2);
    return sum;
}

long sumX2(int** x, int m, int j) {
    long sum = 0;
    for(int i=0; i<m; i++) sum += pow(x[i][j], 2);
    return sum;
}

long sumXY(int** x, int *y, int m, int j) {
    long sum = 0;
    for(int i=0; i<m; i++) sum += x[i][j] * y[i];
    return sum;
}

long sumX(int** x, int m, int j) {
    long sum = 0;
    for(int i=0; i<m; i++) sum += x[i][j];
    return sum;
}

long sumY(int* y, int m) {
    long sum = 0;
    for(int i=0; i<m; i++) sum += y[i];
    return sum;
}

int** generateRandomMatrix(int size) {
    srand(time(NULL)); 
    int** temp = (int**)malloc(sizeof(int*)*size);
    for(int i=0; i<size; i++) {
        temp[i] = (int*)malloc(sizeof(int)*size);
        for (int j=0; j<size; j++) temp[i][j] = rand()%100+1;
    }
    return temp;
}

int* generateRandomY(int size){
    srand(time(NULL)); 
    int* temp = (int*)malloc(sizeof(int)*size);
    for(int i=0; i<size; i++) temp[i] = rand()%100+1;
    return temp;
}

void pearson_cor(int** X, int* y, int m, int n, float* v) {

    long resSumY = sumY(y,m);
    long resSumY2 = sumY2(y,m);
    long resSumY_2 = pow(resSumY,2);

    for(int i=0; i<m; i++) {
        v[i] = (m*sumXY(X,y,m,i)-sumX(X,m,i)*resSumY)/pow((m*sumX2(X,m,i)-pow(sumX(X,m,i),2))*((m*resSumY2)-resSumY_2),0.5);
    }
}

void printMatrix(int **X, int size) {
    for(int i=0; i<size; i++) {
        for (int j=0; j<size; j++) {
            printf("%d ", X[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

int main(int argc, char *argv[]) {
    int size;

    printf("Size: ");
    scanf("%d", &size);

    float* v = (float*)malloc(sizeof(float)*size);
    int** matrix = generateRandomMatrix(size);
    // printMatrix(matrix, size);
    int* y = generateRandomY(size);
    // for(int i=0; i<size; i++) printf("%d ", y[i]);
    printf("\n");

    struct timespec start;
    clock_gettime(CLOCK_MONOTONIC, &start);

    pearson_cor(matrix, y, size, size, v);
    
    struct timespec end;
    clock_gettime(CLOCK_MONOTONIC, &end);
    printf("time: %f seconds\n", (end.tv_sec-start.tv_sec) + (end.tv_nsec-start.tv_nsec) / 1000000000.0); 
    // for(int i=0; i<size; i++) printf("%lf ", v[i]);
    printf("\n");
    return 0;
}